File: android/src/main/java/com/surveysparrow/plugins/ionic/SurveySparrowIonicPluginPlugin.java
package com.surveysparrow.plugins.ionic;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import com.surveysparrow.ss_android_sdk.SsSurvey.CustomParam;

import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;

@CapacitorPlugin(name = "SurveySparrowIonicPlugin")
public class SurveySparrowIonicPluginPlugin extends Plugin {

    private final SurveySparrowIonicPlugin implementation = new SurveySparrowIonicPlugin();

    @PluginMethod
    public void loadFullScreenSurvey(PluginCall call) {
        String domain = call.getString("domain");
        String token = call.getString("token");
        CustomParam[] params = parseParams(call.getObject("params"));
        HashMap<String, String> properties = parseProperties(call.getObject("properties"));

        implementation.loadFullScreenSurvey(domain, token, params, properties, this.getActivity());
        call.resolve();
    }

    @PluginMethod
    public void loadFullScreenSurveyWithValidation(PluginCall call) {
        String domain = call.getString("domain");
        String token = call.getString("token");
        CustomParam[] params = parseParams(call.getObject("params"));
        HashMap<String, String> properties = parseProperties(call.getObject("properties"));

        implementation.loadFullScreenSurveyWithValidation(domain, token, params, properties, this.getActivity());
        call.resolve();
    }

    private CustomParam[] parseParams(JSObject jsParams) {
        if (jsParams == null) {
            return new CustomParam[0];
        }
    
        Iterator<String> keysIterator = jsParams.keys();
        List<CustomParam> paramsList = new ArrayList<>();
    
        while (keysIterator.hasNext()) {
            String key = keysIterator.next();
            String value = jsParams.getString(key);
            paramsList.add(new CustomParam(key, value));
        }

        return paramsList.toArray(new CustomParam[0]);
    }

    private HashMap<String, String> parseProperties(JSObject jsProperties) {
        HashMap<String, String> properties = new HashMap<>();
        if (jsProperties != null) {
            for (Iterator<String> it = jsProperties.keys(); it.hasNext(); ) {
                String key = it.next();
                properties.put(key, jsProperties.getString(key));
            }
        }
        if (!properties.containsKey("langCode")) {
            properties.put("langCode", "en");
        }
        return properties;
    }
}

File: ios/Sources/SurveySparrowIonicPluginPlugin/SurveySparrowIonicPluginPlugin.swift
import Foundation
import Capacitor

@objc(SurveySparrowIonicPluginPlugin)
public class SurveySparrowIonicPluginPlugin: CAPPlugin, CAPBridgedPlugin {
    
    public let identifier = "SurveySparrowIonicPluginPlugin"
    public let jsName = "SurveySparrowIonicPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "loadFullScreenSurvey", returnType: CAPPluginReturnNone),
        CAPPluginMethod(name: "loadFullScreenSurveyWithValidation", returnType: CAPPluginReturnNone)
    ]
    
    private let implementation = SurveySparrowIonicPlugin()
    
    @objc public func loadFullScreenSurvey(_ call: CAPPluginCall) {
        
        let domain = call.getString("domain")
        let token = call.getString("token")
        let params = call.getObject("params") as? [String : String]
        let properties = call.getObject("properties")

        guard let domain = domain, let token = token, let params = params, let properties = properties else {
            call.reject("Invalid or missing parameters")
            return
        }

        DispatchQueue.main.async {
            self.implementation.loadFullScreenSurvey(domain: domain, token: token, params: params, properties: properties)
        }
    
    }
    
    @objc public func loadFullScreenSurveyWithValidation(_ call: CAPPluginCall) {

        let domain = call.getString("domain")
        let token = call.getString("token")
        let params = call.getObject("params") as? [String : String]
        let properties = call.getObject("properties")
        
        guard let domain = domain, let token = token, let params = params, let properties = properties else {
            call.reject("Invalid or missing parameters")
            return
        }

        DispatchQueue.main.async {
            self.implementation.loadFullScreenSurveyWithValidation(domain: domain, token: token, params: params, properties: properties)
        }
    }
}
